### PR TITLE
fix(cross-seed): use correct save path for tracker-named categories in reflink/hardlink mode

### LINKED ISSUE #1703: Bug: Cross-seed creates tracker-named categories with incorrect save paths (reflink/hardlink + by-tracker preset)
## Description

When cross-seed runs with `UseCategoryFromIndexer=true` and reflink/hardlink mode using the `by-tracker` directory preset, newly created categories are assigned incorrect `savePath` values. The category save paths end up pointing to torrent-specific reflink tree directories (e.g., `/data/cross-seed/LST/Movie.Name--0bbfd477`) instead of the tracker-level base directory (e.g., `/data/cross-seed/LST`).

Additionally, when multiple cross-seeds are processed concurrently, a race condition causes category save paths to be shuffled across tracker categories — e.g., the "FearNoPeer" category ends up with LST's path, "LST" gets seedpool's path, etc.

## Steps to reproduce

1. Configure an instance with:
   - Reflinks (or hardlinks) enabled
   - `HardlinkDirPreset` = `by-tracker`
   - `HardlinkBaseDir` includes a cross-seed directory (e.g., `/data/cross-seed/`)
2. Configure cross-seed automation with `UseCategoryFromIndexer = true`
3. Run a dirscan or completion search that matches a torrent across multiple trackers
4. Observe that categories are created with incorrect save paths

## Observed behavior

Categories created by cross-seed have torrent-specific paths:

```json
{
  "FearNoPeer": {
    "name": "FearNoPeer",
    "savePath": "/data/cross-seed/LST/Past.Lives.2023.UHD.BluRay.2160p.DD.5.1.DV.HDR.x265-BHDStudio.mp4--0bbfd477"
  },
  "LST": {
    "name": "LST",
    "savePath": "/data/cross-seed/seedpool (API)/Past.Lives.2023.UHD.BluRay.2160p.DD.5.1.DV.HDR.x265-BHDStudio.mp4--f28fb640"
  },
  "OnlyEncodes+ (API)": {
    "name": "OnlyEncodes+ (API)",
    "savePath": "/data/cross-seed/TorrentLeech/Past.Lives.2023.UHD.BluRay.2160p.DD.5.1.DV.HDR.x265-BHDStudio.mp4--8ec1f765"
  }
}
```

Note the paths are also shuffled across categories (FearNoPeer has LST's path, etc.).

## Expected behavior

Categories should be created with the tracker-level base directory as their save path:

```json
{
  "FearNoPeer": {
    "name": "FearNoPeer",
    "savePath": "/data/cross-seed/FearNoPeer"
  },
  "LST": {
    "name": "LST",
    "savePath": "/data/cross-seed/LST"
  },
  "OnlyEncodes+ (API)": {
    "name": "OnlyEncodes+ (API)",
    "savePath": "/data/cross-seed/OnlyEncodes+ (API)"
  }
}
```

## Root cause

Two issues in `internal/services/crossseed/service.go`:

### 1. Category save path determined too early with wrong fallback

At lines ~4420-4456, `categorySavePath` is determined *before* entering reflink/hardlink processing. The logic is:
- Look up `baseCategory` in qBittorrent's existing categories
- If not found (it won't be — it's a new tracker-named category), fall back to `props.SavePath` (the matched torrent's save path, e.g., `/data/movies`)

This fallback is incorrect for `by-tracker` mode. The correct save path is `{selectedBaseDir}/{trackerName}`, but `selectedBaseDir` isn't known until inside `processReflinkMode()`/`processHardlinkMode()` where `FindMatchingBaseDir()` and `buildHardlinkDestDir()` are called.

### 2. Concurrent processing causes path shuffling

`ensureCrossCategory()` uses `singleflight` and a `createdCategories` sync.Map cache (lines 10015-10026). When multiple cross-seeds for different trackers process concurrently, the first goroutine to reach `ensureCrossCategory` for a given category creates it with whatever `categorySavePath` it computed. Due to the incorrect fallback in issue #1, concurrent goroutines racing through the category creation can associate the wrong tracker's path with the wrong category.

### 3. Existing categories with wrong paths are never corrected

`ensureCrossCategory()` (line 10037-10048) only logs a warning when an existing category has a different save path — it never updates it. So once a category is created with a wrong path, it stays wrong.

## Proposed fix

**Move category creation into the reflink/hardlink processing functions** where the correct path components are already available.

Inside `processReflinkMode()` and `processHardlinkMode()`, after `selectedBaseDir` and `trackerDisplayName`/`incomingTrackerDomain` are resolved (and before the torrent is added), compute the correct category save path:

```go
// For by-tracker preset: category save path is baseDir/trackerName
// For by-instance preset: category save path is baseDir/instanceName  
// For flat preset: category save path is baseDir
categorySavePath := selectedBaseDir
if instance.HardlinkDirPreset == "by-tracker" {
    trackerDisplayName := s.resolveTrackerDisplayName(ctx, incomingTrackerDomain, req)
    categorySavePath = filepath.Join(selectedBaseDir, pathutil.SanitizePathSegment(trackerDisplayName))
} else if instance.HardlinkDirPreset == "by-instance" {
    categorySavePath = filepath.Join(selectedBaseDir, pathutil.SanitizePathSegment(candidate.InstanceName))
}

if crossCategory != "" {
    if err := s.ensureCrossCategory(ctx, candidate.InstanceID, crossCategory, categorySavePath); err != nil {
        // handle error...
    }
}
```

The early category creation block (lines ~4429-4456) should be skipped when reflink or hardlink mode will be used, since those code paths will handle it with the correct save path.

For `ensureCrossCategory()`, keep the existing warn-only behavior for mismatched save paths on existing categories to avoid breaking users' existing setups. The fix prevents incorrect paths from being set going forward.

## Impact

- Categories created with wrong save paths cause confusion in the qBittorrent UI
- If a user subsequently enables AutoTMM on these categories, torrents would be moved to the wrong location
- The shuffled paths make it difficult to understand which tracker's content is where

## Environment

- Instance with local filesystem access, **reflinks enabled**
- `HardlinkDirPreset`: `by-tracker`
- Cross-seed automation with `UseCategoryFromIndexer`: `true`
- Multiple trackers matching the same content concurrently

### PR DESCRIPTION
**Pull request by nitrobass24**
## Summary

- Defer category creation to reflink/hardlink processing functions where the correct base directory and tracker name are resolved
- Compute category save path as `{baseDir}/{trackerName}` instead of falling back to the matched torrent's save path
- Handle fallback to regular mode when reflink/hardlink handlers return `Used=false` — ensure category is created before proceeding
- Track and report category creation failures in hardlink/reflink result messages

## Test plan
- [x] New unit tests for `buildCategorySavePath` covering all presets (by-tracker, by-instance, flat, unknown, empty), edge cases (trailing slash, special characters, no tracker info), and isolation folder exclusion
- [x] Manual test with reflink + by-tracker preset + `UseCategoryFromIndexer=true` across multiple trackers
- [x] Verify categories get correct `{baseDir}/{trackerName}` save paths, not torrent-specific paths
- [x] Verify concurrent cross-seeds don't shuffle save paths across categories

Fixes #1703

## Summary by CodeRabbit

* **Tests**
  * Added comprehensive tests for cross-seed category save-path logic and caching, covering presets, path normalization, tracker display mapping, fallbacks, exclusion of torrent/isolation artifacts, case-insensitive comparisons, and concurrent behavior.

* **Improvements**
  * Category creation now caches and validates normalized effective save paths and revalidates when requested paths change.
  * Link-mode handling avoids inheriting incorrect paths, computes save paths from configured presets, and provides clearer status messages.

### CODE REVIEW

File: internal/services/crossseed/category_savepath_test.go [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			mockStore := &mockTrackerCustomizationStore{
+				customizations: tt.customizations,
+			}
+			if tt.customizations == nil {
+				mockStore.customizations = []*models.TrackerCustomization{}
+			}
```
[the remark this answers is not in the document]
Comment (nitrobass24) [reply internal/services/crossseed/category_savepath_test.go:119]: The mock returns m.customizations directly. resolveTrackerDisplayName passes it to models.ResolveTrackerDisplayName which iterates over the slice with for _, c := range customizations — ranging over nil is safe in Go. So nil would work fine. But this is a trivial
  nitpick — the normalization just avoids a nil slice in the mock, doesn't change any behavior. The test works correctly either way.

File: internal/services/crossseed/service.go [hunk 2]
```diff
@@ -4488,6 +4494,20 @@ func (s *Service) processCrossSeedCandidate(
...
+	if crossCategory != "" && (useReflinkMode || useHardlinkMode) {
+		if err := s.ensureCrossCategory(ctx, candidate.InstanceID, crossCategory, categorySavePath); err != nil {
+			log.Warn().Err(err).
+				Str("category", crossCategory).
+				Str("savePath", categorySavePath).
+				Msg("[CROSSSEED] Failed to ensure category exists on fallback to regular mode, continuing without category")
+			crossCategory = ""
+			categoryCreationFailed = true
+		}
+	}
```
[the remark this answers is not in the document]
Comment (nitrobass24) [reply internal/services/crossseed/service.go]: All notUsed returns are at lines 10990-11009 and 11581-11600 — all before ensureCrossCategory is called (which is at ~11100 for hardlink, ~11700 for reflink). Once the code reaches ensureCrossCategory, it never returns notUsed. It either succeeds and returns Used=true, or fails and returns Used=true with an error result.
The concern is not valid — ensureCrossCategory and createdCategories are never populated before a Used=false return. The fallback code at line 4500 is correct.

[the remark this answers is not in the document]
Comment (nitrobass24) [reply internal/services/crossseed/service.go]: When FallbackToRegularMode is enabled, handleError returns notUsed, not Used: true. That means ensureCrossCategory can fire inside the link mode function (creating the category with a link-mode path), and then handleError returns notUsed on a downstream failure like BuildPlan or AddTorrent, falling through to regular mode with a mismatched category path.

Applied Option A — the fallback block now skips category isolation instead of retrying ensureCrossCategory. This avoids inheriting a link-mode save path in regular mode.

File: internal/services/crossseed/service.go [hunk 3]
```diff
@@ -11077,6 +11095,21 @@ func (s *Service) processHardlinkMode(
...
+		categorySavePath := s.buildCategorySavePath(ctx, instance, selectedBaseDir, incomingTrackerDomain, candidate, req)
+		if err := s.ensureCrossCategory(ctx, candidate.InstanceID, crossCategory, categorySavePath); err != nil {
+			log.Warn().Err(err).
+				Str("category", crossCategory).
+				Str("savePath", categorySavePath).
+				Msg("[CROSSSEED] Hardlink mode: failed to ensure category exists, continuing without category")
+			crossCategory = ""
+			categoryCreationFailed = true
+		}
+	}
```
[the remark this answers is not in the document]
Comment (nitrobass24) [reply internal/services/crossseed/service.go:11168]: 1. This is pre-existing behavior, not introduced by our PR. The warn-and-cache logic existed before our changes.                                                                                                                                                           
2. It's intentional — the original issue (#1703) explicitly noted: "keep the existing warn-only behavior for mismatched save paths on existing categories to avoid breaking users' existing setups."

**Comment by nitrobass24** [comment]
@s0up4200 This one is ready for your review.
